homeassistant/components/synology_dsm/backup.py

```diff
@@ -0,0 +1,217 @@
+from homeassistant.helpers.aiohttp_client import ChunkAsyncStreamIterator
+from homeassistant.helpers.json import json_dumps
+from homeassistant.util.json import JsonObjectType, json_loads_object
+
+from .const import (
+    CONF_BACKUP_PATH,
+    CONF_BACKUP_SHARE,
+    DOMAIN,
+    SYNOLOGY_DATA_BACKUP_AGENT_LISTENERS,
+)
+from .models import SynologyDSMData
+
+LOGGER = logging.getLogger(__name__)
+
+
+async def async_get_backup_agents(
+    hass: HomeAssistant,
+) -> list[BackupAgent]:
+    """Return a list of backup agents."""
+    if not (
+        entries := hass.config_entries.async_loaded_entries(DOMAIN)
+    ) or not hass.data.get(DOMAIN):
+        LOGGER.debug("No proper config entry found")
+        return []
+    agents: list[BackupAgent] = []
+    for entry in entries:
+        syno_data: SynologyDSMData = hass.data[DOMAIN][entry.unique_id]
+        if syno_data.api.file_station and entry.options.get(CONF_BACKUP_PATH):
+            agents.append(SynologyDSMBackupAgent(hass, entry))
+    return agents
+
+
+@callback
+def async_register_backup_agents_listener(
+    hass: HomeAssistant,
+    *,
+    listener: Callable[[], None],
+    **kwargs: Any,
+) -> Callable[[], None]:
+    """Register a listener to be called when agents are added or removed.
+
+    :return: A function to unregister the listener.
+    """
+    hass.data.setdefault(SYNOLOGY_DATA_BACKUP_AGENT_LISTENERS, []).append(listener)
+
+    @callback
+    def remove_listener() -> None:
+        """Remove the listener."""
+        hass.data[SYNOLOGY_DATA_BACKUP_AGENT_LISTENERS].remove(listener)
+
+    return remove_listener
+
+
+class SynologyDSMBackupAgent(BackupAgent):
+    """Synology DSM backup agent."""
+
+    domain = DOMAIN
+
+    def __init__(self, hass: HomeAssistant, entry: ConfigEntry) -> None:
+        """Initialize the Synology DSM backup agent."""
+        super().__init__()
+        LOGGER.debug("Initializing Synology DSM backup agent for %s", entry.unique_id)
+        self.name = entry.title
+        self.path = (
+            f"{entry.options[CONF_BACKUP_SHARE]}/{entry.options[CONF_BACKUP_PATH]}"
+        )
+        syno_data: SynologyDSMData = hass.data[DOMAIN][entry.unique_id]
+        self.api = syno_data.api
+
+    @property
+    def _file_station(self) -> SynoFileStation:
+        if not self.api.file_station:
+            raise BackupAgentError("Synology FileStation API not available")
+        return self.api.file_station
+
+    async def async_download_backup(
+        self,
+        backup_id: str,
+        **kwargs: Any,
+    ) -> AsyncIterator[bytes]:
+        """Download a backup file.
+
+        :param backup_id: The ID of the backup that was returned in async_list_backups.
+        :return: An async iterator that yields bytes.
+        """
+        if not await self.async_get_backup(backup_id):
+            raise BackupAgentError("Backup not found")
+
+        try:
+            resp = await self._file_station.download_file(
```

Comment:
`iter_chunks` is usually a bit faster since it doesn't have to do as much copy on the data and we aren't worried about the source sending oversized chunks

Reply:
we are using the [ChunkAsyncStreamIterator](https://github.com/home-assistant/core/blob/005ae3ace66b7661042c472d82e98dd161d7caf1/homeassistant/helpers/aiohttp_client.py#L86) the read the data from the stream - or do you mean something different?

Reply:
https://github.com/mib1185/py-synologydsm-api/blob/7e4d5f983bacc4118683da910996cb51bf8bba21/src/synology_dsm/api/file_station/__init__.py#L155

It should be a bit faster to change that line to 

`async for data, _ in result.content.iter_chunks():`

Reply:
But there is another issue as well since its doing the blocking I/O in the event loop
https://github.com/mib1185/py-synologydsm-api/blob/7e4d5f983bacc4118683da910996cb51bf8bba21/src/synology_dsm/api/file_station/__init__.py#L154

blocking open https://github.com/mib1185/py-synologydsm-api/blob/7e4d5f983bacc4118683da910996cb51bf8bba21/src/synology_dsm/api/file_station/__init__.py#L154
blocking write https://github.com/mib1185/py-synologydsm-api/blob/7e4d5f983bacc4118683da910996cb51bf8bba21/src/synology_dsm/api/file_station/__init__.py#L156

Reply:
ok ... i will fix this in the lib, but anyway we don't use this part of the lib code here in the backup-agent, so i don't think this is a blocker for this PR